Added changes from feedback call (#4)
sreeram-venkitesh/enhancements: keps/sig-api-machinery/4595-cel-crd-additionalprintercolumns/README.md

	return reflectSlice, nil
}

func (c ColumnCompilationResult) PrintResults(w io.Writer, results []reflect.Value) error {
	for _, result := range results {
		klog.V(1).Info("Inside FindResults for loop")
		klog.V(1).Info(result)
		var str string
		switch result.Kind() {
		case reflect.String:
			str = result.String()
		case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
			str = fmt.Sprintf("%d", result.Int())
		case reflect.Float32, reflect.Float64:
			str = fmt.Sprintf("%f", result.Float())
		case reflect.Bool:
			str = fmt.Sprintf("%v", result.Bool())
		default:
			str = fmt.Sprintf("%v", result.Interface())
		}

		_, err := w.Write([]byte(str))
		if err != nil {
			klog.V(1).Info("Error inside cel printresults")
			klog.V(1).Info(err)
			return err
		}
	}

	return nil
}
```

With this, the current `ConvertToTable` method can support additionalPrinterColumns without any changes.

### Cost Calculation and benchmarking performance with JSONPath

> A big part of the discussions for our proposal was the CEL cost limits since this is the first time CEL is added to the read path. As part of this we've done some benchmarking of the time it takes to parse and compile equivalent JSONPath and CEL expressions.

Reviewer: :+1:

Alpha:

TODO: Mention which files we'll edit

> staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation_test.go

Reviewer: ```suggestion
[staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation_test.go](https://github.com/kubernetes/kubernetes/blob/e54719bb6674fac228671e0786d19c2cf27b08a3/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation_test.go)
```

- Test that validation passes when we create an additionalPrinterColumn with an expression field with valid CEL expression
- Test that validation fails when we create an additionalPrinterColumn with an expression field with an invalid CEL expression
- Test that existing behaviour of jsonPath is not altered when creating CRDs with only jsonPath additionalPrinterColumns
- Test that validation fails when we create an additionalPrinterColumn with both jsonPath and expression fields
- Test that validation passes when we create multiple additionalPrinterColumns with both jsonPath and expression fields
- Test that validation fails when we try to create an additionalPrinterColumn with expression field when the feature gate is turned off

> staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/tableconvertor/tableconvertor_test.go

Reviewer: ```suggestion
[staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/tableconvertor/tableconvertor_test.go](https://github.com/kubernetes/kubernetes/blob/bc302fa4144d21a338683cd83701661f97be4aba/staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/tableconvertor/tableconvertor_test.go)
```

For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
https://storage.googleapis.com/k8s-triage/index.html
-->

> test/integration/apiserver/crd_additional_printer_columns_test.go

Reviewer: ```suggestion
[test/integration/apiserver/crd_additional_printer_columns_test.go](https://github.com/kubernetes/kubernetes/tree/bc302fa4144d21a338683cd83701661f97be4aba/test/integration/apiserver)
```

We expect no non-infra related flakes in the last month as a GA graduation criteria.
-->

<!-- - <test>: <link to test coverage> -->

> We will test the edge cases mostly in integration test and unit test. If needed, we can add e2e tests before beta graduation.

Reviewer: what edge cases? and `before beta graduation` means not alpha right? it means when we target beta graduation?

```suggestion
We will test the edge cases mostly in integration test and unit test. If needed, we can add e2e tests while planning for beta graduation.
```

- Gather feedback from developers and surveys
- Add e2e tests
- Add appropriate metrics for additionalPrinterColumns usage and CEL cost usage
- More benchmarking to compare JSONPath and CEL execution and modify CEL cost if needed

#### GA

- N examples of real-world usage

> - N installs

Reviewer: N installs

How we will come up with stats for this?

N examples of real-world usage makes proper sense to me since we already know few areas which will immediately start using this feature once available.

  cluster required to make on upgrade, in order to make use of the enhancement?
-->

<!-- No change in how users upgrade/downgrade their clusters. This feature may remove -->
<!-- complexity by removing risk that a tightened validation on Kubernetes' part -->
<!-- does not break workflow. -->

No changes are required for a cluster to make an upgrade and maintain existing behavior. There is new API that does not effect the cluster during upgrade. It only has effects if it is used after the upgrade.

> If a cluster is downgraded, no changes are required. The cluster continues to work as expected since the alpha version will have functionality compatible with beta and stable release, any downgrade will be to a version that also contains the feature.

Reviewer: This wording still doesn't make sense to me sadly :(

  CRI or CNI may require updating that component before the kubelet.
-->

This feature is implemented in the kube-apiserver component, skew with other kubernetes components do not require coordinated behavior.

> Clients should ensure the kube-apiserver is fully rolled out before using the feature.

Reviewer: :+1:

NOTE: Also set `disable-supported` to `true` or `false` in `kep.yaml`.
-->

> Yes, if the feature is disabled after being used, the existing additionalPrinterColumns with JSONPaths would work as expected. Existing resources with CEL expressions in their additionalPrinterColumn definition would be ignored and those columns will not be printed if the feature is disabled.

Reviewer: minor typo?

```suggestion
Yes, if the feature is disabled after being used, the existing additionalPrinterColumns with JSONPath would work as expected. Existing resources with CEL expressions in their additionalPrinterColumn definition would be ignored and those columns will not be printed if the feature is disabled.
```

###### What happens if we reenable the feature if it was previously rolled back?

> CRDs which had failed validation previously might now succeed if the CEL expression is valid. Existing CRDs additionalPrinterColumns defined with CEL expression would start working again after the feature has been reenabled.

Reviewer: :+1:

https://github.com/kubernetes/kubernetes/pull/97058/files#diff-7826f7adbc1996a05ab52e3f5f02429e94b68ce6bce0dc534d1be636154fded3R246-R282
-->

> We have unit and integration tests to make sure that the feature enablement and disablement works as intended.

Reviewer: I thought we can still add "will have" instead of "have" :sweat_smile: 

```suggestion
We will have unit and integration tests to make sure that the feature enablement and disablement works as intended.
```

are missing a bunch of machinery and tooling and can't do that now.
-->

> We're planning to test upgrade-> downgrade -> upgrades before graduating to beta.

Reviewer: :+1:

logs or events for this purpose.
-->

> The cluster admin can check if the CRDAdditionalPrinterColumnCEL feature gate is turned on. If yes, the admin can check if any CRD has additionalPrinterColumns which are defined with `expression` instead of `jsonPath`.

Reviewer: remove the instead of `jsonPath` part?

```suggestion
The cluster admin can check if the CRDAdditionalPrinterColumnCEL feature gate is turned on.
If yes, the admin can further check if any CRD has any columns defined under `additionalPrinterColumns` section which are using the new `expression` field.
```

[supported limits]: https://git.k8s.io/community//sig-scalability/configs-and-limits/thresholds.md
-->

> Since the CEL expressions are compiled and evaluated in the kube-apiserver, depending on the complexity of the CRDs and the expressions defined, we may see a non-negligible increase of CPU usage. We are planning to benchmark this before beta graduation.

Reviewer: :+1:
